mission/mission_handler_registry.hpp

```diff
@@ -0,0 +1,102 @@
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ ******************************************************************************/
+
+/*******************************************************************************
+ * \file mission_handler_registry.hpp
+ *
+ * \author MAV'RIC Team
+ * \author Matthew Douglas
+ *
+ * \brief The mission handler registry is responsible for registering and
+ *        obtaining mission handlers based on an inputted waypoint
+ *
+ ******************************************************************************/
+
+
+#ifndef MISSION_HANDLER_REGISTRY__
+#define MISSION_HANDLER_REGISTRY__
+
+class Mission_handler;
+
+#include "mission/waypoint.hpp"
+
+#define MAX_REGISTERED_MISSION_HANDLERS 20
+
+/*
+ * N.B.: Reference Frames and MAV_CMD_NAV are defined in "maveric.h"
+ */
+
+class Mission_handler_registry
+{
+public:
+    Mission_handler_registry();
+
+    /**
+     * \brief   Registers the inputted handler to the array of known
+     *          mission handlers. Performs a check to see if that
+     *          object is already within the array.
+     *
+     * \param   handler     The new mission handler
+     *
+     * \return  Success
+     */
+    bool register_mission_handler(Mission_handler& handler);
+
+    /**
+     * \brief   Attempts to get the mission handler form the registry.
+     *
+     * \details If there is a mission handler that can take the waypoint,
+     *          then the registry will return the first mission handler
+     *          that can take it. If no mission handler can take the
+     *          waypoint, then the mission handler will return NULL.
+     *
+     * \param   waypoint    The waypoint to be handled
+     *
+     * \return  The first acceptable mission handler or NULL
+     */
+    Mission_handler* get_mission_handler(const Waypoint& waypoint);
+
+private:
+    /**
+     * Array of registered mission handlers. Other classes will
+     * register a mission handler through the function
+     * Mission_planner::register_mission_handler(Mission_handler*).
+     */
+    Mission_handler* registered_mission_handlers_[MAX_REGISTERED_MISSION_HANDLERS];
```

Comment:
Here we could use the same trick as in Scheduler, Data_logging, Onboard_parameters, Periodic_telemetry and Message_handler to make the max number of handler customizable.